src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorParser.cs

```diff
@@ -58,43 +61,58 @@ public class MvcRazorParser : RazorParser
             var descriptors = new List<TagHelperDirectiveDescriptor>();
 
             // For tag helpers, the @removeTagHelper only applies tag helpers that were added prior to it.
-            // Consequently we must visit tag helpers outside-in - furthest _ViewStart first and nearest one last. This
+            // Consequently we must visit tag helpers outside-in - furthest _GlobalImport first and nearest one last. This
             // is different from the behavior of chunk merging where we visit the nearest one first and ignore chunks
             // that were previously visited.
-            var chunksFromViewStarts = inheritedCodeTrees.Reverse()
-                                                         .SelectMany(tree => tree.Chunks);
-            var chunksInOrder = defaultInheritedChunks.Concat(chunksFromViewStarts);
+            var chunksFromGlobalImports = inheritedCodeTrees
+                .Reverse()
+                .SelectMany(tree => tree.Chunks);
+            var chunksInOrder = defaultInheritedChunks.Concat(chunksFromGlobalImports);
             foreach (var chunk in chunksInOrder)
             {
-                var addHelperChunk = chunk as AddTagHelperChunk;
-                if (addHelperChunk != null)
+                if (chunk is AddTagHelperChunk)
                 {
-                    var descriptor = new TagHelperDirectiveDescriptor(addHelperChunk.LookupText,
-                                                                      SourceLocation.Undefined,
-                                                                      TagHelperDirectiveType.AddTagHelper);
+                    var descriptor = CreateInheritedTagHelperDirectiveDescriptor(
+                        ((AddTagHelperChunk)chunk).LookupText,
+                        TagHelperDirectiveType.AddTagHelper);
+
                     descriptors.Add(descriptor);
                 }
-                else
+                else if (chunk is RemoveTagHelperChunk)
```

Comment:
nit: `continue;` at the bottom of the previous block and at the end of this one.  then can do all of the `chunk as ...` casts at the same level without doing them all at once.